pennylane/devices/tf_ops.py

```diff
@@ -191,3 +191,103 @@ def CRot(a, b, c):
         :math:`|0\rangle\langle 0|\otimes \mathbb{I}+|1\rangle\langle 1|\otimes R(a,b,c)`
     """
     return tf.linalg.diag(CRZ(c)) @ (CRY(b) @ tf.linalg.diag(CRZ(a)))
+
+
+def DoubleExcitation(phi):
+    r"""
+    Double excitation rotation.
+    Args:
+        phi (float): rotation angle
+    Returns:
+        tf.Tensor[complex]: Double excitation rotation matrix
+    """
+    phi = tf.cast(phi, dtype=C_DTYPE)
+    c = tf.cos(phi / 2)
+    s = tf.sin(phi / 2)
+
+    U = [[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+     [0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+     [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+     [0, 0, 0, c, 0, 0, 0, 0, 0, 0, 0, 0, -s, 0, 0, 0],
+     [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+     [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+     [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+     [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
+     [0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
+     [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
+     [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
+     [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
+     [0, 0, 0, s, 0, 0, 0, 0, 0, 0, 0, 0, c, 0, 0, 0],
+     [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
+     [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
+     [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]]

```

Comment:
is this because tensor indexing doesn't work in tensorflow? You could try using [`tf.scatter_nd`](https://www.tensorflow.org/api_docs/python/tf/scatter_nd) to avoid writing out the full matrix.

Reply:
Thanks Josh! I took a look and it's still pretty cumbersome to write down the matrix using `tf.scatter_nd` . As far as I can tell the only way to use `tf.scatter` is to populate an entire tensor index at a time for a tensor of all zeros. In this case it means populating entire rows. This would still be helpful if some rows where all zeros, but instead it seems we have to specify each full row directly, which is as difficult as writing down the entire matrix. For example, to build the 2x2 identity we'd have to do something like:

```python

indices = tf.constant([[0], [1]])
updates = tf.constant([[1, 0], [0, 1]])
shape = tf.constant([2, 2])
scatter = tf.scatter_nd(indices, updates, shape)

print(scatter)

>>> tf.Tensor(
[[1 0]
 [0 1]], shape=(2, 2), dtype=int32)
```

Reply:
Ah. To be honest I don't 100% understand `scatter_nd` 😆